### PR TITLE
Simple (slow!) instruction synthesis + pass support

### PR DESCRIPTION
**Pull request by regehr**
I think the interesting part of this patch is the getValue() function in the Pass

### CODE REVIEW

File: lib/Extractor/Candidates.cpp [hunk 1]
```diff
@@ -354,6 +354,9 @@ Inst *ExprBuilder::build(Value *V) {
       ; // fallthrough to return below
     }
   } else if (auto GEP = dyn_cast<GetElementPtrInst>(V)) {
+
+    return makeArrayRead(V);
```
Comment (pcc) [inline lib/Extractor/Candidates.cpp]: The dead code below (and `buildGEP`) should be removed. When we do support GEP as a native instruction, it seems  that `buildGEP` would look very different.

File: lib/Extractor/Solver.cpp [hunk 2]
```diff
@@ -50,27 +62,122 @@ class BaseSolver : public Solver {
...
+              { IC.getConst(APInt(I->Width, -1)), I }));
+          Guesses.push_back(IC.getInst(Inst::Sub, LHS->Width,
+              { IC.getConst(APInt(I->Width, 0)), I }));
+          if (I->Width == 16 || I->Width == 32 || I->Width == 64)
+            Guesses.push_back(IC.getInst(Inst::BSwap, LHS->Width, {I}));
+          if (I->Width == 8 || I->Width == 16 || I->Width == 32 ||
+              I->Width == 64 || I->Width == 256) {
+            Guesses.push_back(IC.getInst(Inst::CtPop, LHS->Width, {I}));
+            Guesses.push_back(IC.getInst(Inst::Cttz, LHS->Width, {I}));
+            Guesses.push_back(IC.getInst(Inst::Ctlz, LHS->Width, {I}));
```
Comment (pcc) [inline lib/Extractor/Solver.cpp]: Are bswap/ctpop/cttz/ctlz likely enough to be worth guessing every time?

Comment (regehr) [inline lib/Extractor/Solver.cpp]: Definitely not, but keep in mind that this code only runs when the user asks for it and also it's intended to be provisional and will get ripped out when we have a real synthesis strategy.

File: lib/Extractor/Candidates.cpp (hunk 1 above)
Comment (regehr) [inline lib/Extractor/Candidates.cpp]: OK I'll pull out the current GEP code and we can replace it later.

**Comment by regehr** [comment]
I'll bring this back after Raimondas's synthesis patch lands

**Pull request closed without merging**